### PR TITLE
Added stacked what where autoencoder.

### PR DESCRIPTION
**Pull request by antonmbk**
An example script that trains a SWWAE powered by residual blocks on MNIST. Trains fast. Creates very good reconstructions (of MNIST).

I am new to PR's here on Github but need to start somewhere.  I have written this using code that is completely agnostic to backend, however, I dont have Tensorflow so cannot test with it.  Maybe people can help with that?  

Otherwise, someone can take the code and submit it themselves, if I have made too many errors in this PR.  The original code on its own is here, along with a nice image that the script generates:
https://github.com/antonmbk/Salami/blob/master/mnist_swwae.py

-Anton

### CODE REVIEW

**Comment by fchollet** [comment]
The good:

- it's an interesting example
- it's fairly short

The bad: 

- the style doesn't follow PEP8 / Keras conventions
- there are no explanations, docs or even pointers to explanations. Few people know what a SWWAE is.
- some coding patterns should be revised (again, that's mostly style).

Please revise the style and add a detailed docstring, as well as more comments.

**Comment by antonmbk** [comment]
With regard to the bad:
1)  Will hop on this ASAP.  If anyone less busy than @fchollet wants to help me with this, I admit I'm not an expert.
2) Shame on me,  I get lost on the fact that many people still don't know about the Pooling "Switch" reconstruction methods.  But to not even point to the SWWAE, Deconvnet, or Res Net papers is silly, so will add that and a better explanation soon.
3) I can guess what these issues are, but let me focus on 1 and 2 before taking up anyone else's time.

**Comment by antonmbk** [comment]
With regard to the bad:
1) I didn't know there was a PEP8 in Anaconda distribution that let me easily make the checks and correct them, so I did that and this is fixed yay. 
2) I have included ~45 lines at the beginning describing what I think are the two big exemplified ideas, and provided 4 informative references.
3) If we can spot more issues (or things that irk anyone on Keras dev team), lmk so I can make proper adjustments.

Thanks,
Anton

**Comment by fchollet** [comment]
Much better now. Please use a PEP8 linter to get of rid of the remaining issues (e.g. missing spaces around operators).

The docstring and code look good to me.

File: examples/mnist_swwae.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+'''Trains a stacked what-where autoencoder built on residual blocks on the
+MNIST dataset.  It exemplifies two influential methods that have been developed
+the past few years.
```
Comment (farizrahman4u) [inline examples/mnist_swwae.py]: developed **in** the

**Comment by antonmbk** [comment]
> Please use a PEP8 linter to get of rid of the remaining issues (e.g. missing spaces around operators).

I used two PEP8 linters (pycodestyle and pep8) and neither brought up E221 through E228.  Is there a certain mode or option I need to run the linter with?  All they brought up was the import issue because the imports came after the numpy seed.  Otherwise, I added spaces around + and -, when they are used a binary operators based on your hint.  If there are any tips on why these two PEP8 linters did not catch the operator space issues, please let me know, because there may be more lurking about in the code (although I tried my best to find them myself).

File: examples/mnist_swwae.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+# Shape of input to train on (note that model is fully convolutional however)
+input_shape = X_train.shape[1:]
+# The final list of the size of axis=1 for all layers, including input
+nfeats_all = [input_shape[0]] + nfeats
+
+# First build the encoder, all the while keeping track of the "where" masks
+img_input = Input(shape=input_shape)
+
+# We push the "where" masks to the following list
+wheres = [None]*nlayers
```
Comment (fchollet) [inline examples/mnist_swwae.py]: PEP8

File: examples/mnist_swwae.py (hunk 1 above)
Comment (fchollet) [inline examples/mnist_swwae.py]: Please fix.

**Comment by fchollet** [comment]
LGTM, thanks!
